providers/github_test.go

```diff
@@ -220,12 +220,12 @@ func TestGitHubProviderGetEmailAddressWithNoAccessToPrivateRepo(t *testing.T) {
 	p.SetRepo("oauth2-proxy/oauth2-proxy", "")
 
 	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.NotEqual(t, nil, err)
-	assert.Equal(t, "", email)
+	err := p.getEmail(context.Background(), session)
+	assert.NoError(t, err)
```

Comment:
This test was broken before -- The error catch was catching an error in the `testGitHubBackend` by having an empty `[]string{}` for `/repo/oauth2-proxy/oauth2-proxy`.

I disagree with the underlying behavior of `hasRepo` -- `getEmail` raises an error and doesn't set an Email on an error in this method or a false return.  In the `false` case, the email that used to be returned was `""` and `nil` error was raised.

This refactor now at least makes it so `""` won't be returned and clobber a `session.Email` that might already be set. But I wonder if it is best to return an actual error? (I imagine this worked in the past to restrict access accidentally by setting the `session.Email` to "" which failed the email domain validation.

Reply:
To clarify, if `hasRepo` fails, ie the user has failed the authorization check, what happens presently?

Reply:
https://github.com/oauth2-proxy/oauth2-proxy/blob/8be97f25e7276d2d46f002199f9eda345fc0800d/providers/github.go#L402

It looks like it will return `"", err` is either `hasRepo` errors or returns `false`.  This is fine at the moment, but not great -- and got accidental security from from our email validator not allowing empty emails.  But that will change with this PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/770

Reply:
Even though it goes beyond  a refactor, I'm thinking we need to make `hasRepo` returning false return an error so the security logic is completely contained here appropriately.

Reply:
This pattern is all over the place in this provider:

```
err != nil || !ok {
  return err
}
```

is all over the place. A lot of `nil` errors are potentially returned in auth codepaths.  Punting for now, but this will need to be refactored later once we have the better `Authorization` scheme.